net: shell: ensure the shell `sh` is valid before call shell_printf

ndrs-pst opened the pull request:
- It is possible that the `sh` was not set before use. This change adds a NULL check for `sh` in the following macros: PR, PR_SHELL, PR_ERROR, PR_INFO, and PR_WARNING.
Fixes #68793.

- Since PR, PR_SHELL, PR_ERROR, PR_INFO, and PR_WARNING already have
an embedded `sh` NULL check, we can remove the change from PR #68809.

ndrs-pst committed:
net: shell: ensure the shell `sh` is valid before call shell_printf

It is possible that the `sh` was not set before use.
This change adds a NULL check for `sh` in the following macros:
PR, PR_SHELL, PR_ERROR, PR_INFO, and PR_WARNING.
In case `sh` is NULL, the above macros will call `printk` instead.

Fixes #68793

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>

rlubos commented on subsys/net/lib/shell/net_shell_private.h:
```diff
@@ -8,20 +8,40 @@
 
-#define PR(fmt, ...)						\
-	shell_fprintf(sh, SHELL_NORMAL, fmt, ##__VA_ARGS__)
+#define PR(fmt, ...)					\
```
Either keep all of the backslashes aligned within the macro, or just drop the extra backslash indentation. I don't see a point of having some of them aligned, and some not.

ndrs-pst replied:
Then I just let `.clang-format` decide for me 😄

rlubos reviewed (commented):
Looks ok, just some formatting nit.

jukkar commented on subsys/net/lib/shell/net_shell_private.h:
```diff
@@ -8,20 +8,40 @@
 
-#define PR(fmt, ...)						\
-	shell_fprintf(sh, SHELL_NORMAL, fmt, ##__VA_ARGS__)
+#define PR(fmt, ...)                                                                               \
```
Why do we have such a long lines, these are now max 100 chars long. Please adjust the length of the lines to be much smaller.

EricNRS commented on subsys/net/lib/shell/net_shell_private.h:
```diff
@@ -8,20 +8,40 @@
+	do {                                                                                       \
+		if (sh) {                                                                          \
+			shell_fprintf(sh, SHELL_NORMAL, fmt, ##__VA_ARGS__);                       \
+		}                                                                                  \
```
For the else case, please add `printk(fmt, ##__VA_ARGS__);` per option 2 in the issue discussion (this is an update from the initial discussion).

Same applies to the other PR* macros in the rest of the commit.

jukkar replied:
There is way too much back and forth with this simple PR and it seems to be stuck atm. In order to proceed lets add the `printk()` as @EricNRS  suggested above and lets beautify the `\` as I suggested so that there is not so much empty space and the line is shorter. This is max. 10 min work and then we can move on to more difficult tasks.

ndrs-pst replied:
Following @jukkar's strong recommendation to incorporate a check for `sh being NULL` and to fallback to `printk` within the `PR_*` macro, rather than exploring alternative options, I've decided to proceed with this approach. While I had hoped for a different solution, I recognize the importance of moving forward on this matter.

---
Ping @EricNRS / @trantanen if you have any response about this approach.

EricNRS replied:
Yes, this is fine.  It is just restoring it back to the original version before the cleanup.

trantanen replied:
I'd probably add the setting of shell pointer as in https://github.com/zephyrproject-rtos/zephyr/pull/68812#issuecomment-1959466786 to make it work better in many cases.
Memory penalty for the printk seems to be a little high...

But if you need to keep it as it is, then that's it 😄

ndrs-pst replied:
@trantanen, I share your views and am keen to see your proposed solutions in another PR.
Looking forward to it! Regards.

ndrs-pst commented:
FWIW, I built **"samples/net/wifi"** on the **"disco_l475_iot1"** board to compare the code footprint of `wifi_shell.c` among different options.

| FOOTPRINT  | ORIGINAL | OPTION1 | OPTION2 | OPTION3 (BELOW) |
| ---------- | -------- | ------- | ------- | ------- |
| `wifi_shell.c`  | 8,504 B | 9,008 B | 10,780 B | 8,400 B |

---
Now, I have added another call to `OPTION3`, as I proposed in https://github.com/zephyrproject-rtos/zephyr/issues/68793

```
#if IS_ENABLED(CONFIG_SHELL_BACKEND_SERIAL)
	context.sh = shell_backend_uart_get_ptr();
#elif IS_ENABLED(CONFIG_LOG_BACKEND_RTT)
	context.sh = shell_backend_rtt_get_ptr();
#else
	BUILD_ASSERT(0, "Unsupported shell backend");
#endif
```

Then, I was able to remove several lines of code that were initially trying to assign context.sh = sh.

**Pros**: This allows the use of PR macros without a `NULL` check and does not affect other files in `subsys/net/lib/shell`.
**Cons**: It limits the option of the backend to just `SERIAL/RTT`, and later, if needed, we can add `MQTT/TELNET`.

EricNRS commented:
> **Pros**: This allows the use of PR macros without a `NULL` check and does not affect other files in `subsys/net/lib/shell`. 
> **Cons**: It limits the option of the backend to just `SERIAL/RTT`, and later, if needed, we can add `MQTT/TELNET`.

Would that break log capture backends like Golioth?



Interesting footprint comparisons -- I didn't fully realise the amount of extra instructions for the extra call.  I am starting to wonder if the NULL check and `printk()` belongs in `shell_vfprintf()`.  Replace the `__ASSERT_NO_MSG(sh);` with the `if (!sh) printk` code.

https://github.com/zephyrproject-rtos/zephyr/blob/40b53d5d774705d27183aaa1e42f2f322250da10/subsys/shell/shell.c#L1507-L1517

ndrs-pst commented:
@EricNRS, I just realized the snippet was wrong it should be `CONFIG_SHELL_BACKEND_RTT` instead of `CONFIG_LOG_BACKEND_RTT`. So, I think this change would not relate to `CONFIG_LOG_*`

ndrs-pst commented:
I've just discovered another PR (https://github.com/zephyrproject-rtos/zephyr/pull/68809) related to this fix that's already merged.
Now, this begs the question: what is the actual solution for this matter? 😄

EricNRS commented:
Their fix is equivalent to your original upload which checks for null, but discards the message.  I find the messages useful in automated runs where the serial output is captured, but no commands are entered.

ndrs-pst commented:
@jukkar, do you have any comments on the couple of discussions since your last requested change? 
According to the PR (https://github.com/zephyrproject-rtos/zephyr/pull/68809) which has been merged, yet this issue remains open.
I'm just concerned that this may become a roadblock to the 3.6.0 release. (since it was added to the v3.6.0 milestone)

jukkar commented:
> do you have any comments on the couple of discussions since your last requested change?

I think we can have this change, I was just questioning the macro formatting, the longs lines with `\` just look bad, if you fix that we can proceed with this one.

ndrs-pst commented:
@jukkar, thank you for your reply.
By the way, don't you think the PR (https://github.com/zephyrproject-rtos/zephyr/pull/68809) already addresses https://github.com/zephyrproject-rtos/zephyr/issues/68793, but just hinders @EricNRS's use case.

What about `OPTION3`, since you haven't mentioned it yet?
Perhaps you have some insights on why we shouldn't go with this option.

jukkar commented:
> By the way, don't you think the PR (#68809) already addresses #68793, but just hinders @EricNRS's use case.

I have never seen the crash with shell, but if the already merged #68809 solves the problem then this PR is just adding extra checks. So I suppose we would need confirmation from @EricNRS whether this PR can be omitted.

jukkar commented:
> What about `OPTION3`, since you haven't mentioned it yet?

I think the cons you mention i.e., `It limits the option of the backend to just SERIAL/RTT, and later, if needed, we can add MQTT/TELNET.` means we should first think other solutions.

EricNRS commented:
> > By the way, don't you think the PR (#68809) already addresses #68793, but just hinders @EricNRS's use case.
> 
> I have never seen the crash with shell, but if the already merged #68809 solves the problem then this PR is just adding extra checks. So I suppose we would need confirmation from @EricNRS whether this PR can be omitted.

PR [#68809](https://github.com/zephyrproject-rtos/zephyr/pull/68809) fixes the crash, but it drops the notification messages, so it does change the existing behaviour.

Based upon binary code size increases detailed by @ndrs-pst due to the extra calls when adding the `if (sh)` check to the `PR_*` macros, I wonder if the better solution is to move the check into a new net_shell_fprintf() function that does the `if (sh)` check and calls shell_fprintf or printk as appropriate and remove the chang from PR [#68809](https://github.com/zephyrproject-rtos/zephyr/pull/68809).  This would preserve exisiting behaviour for anyone using it and it accomplishes the cleanup and code size reduction that @ndrs-pst was working on.

I am away from my computer for the next 4 days, but let me know if anything is unclear and I will reply when I am back.

ndrs-pst commented:
@jukkar Based on @EricNRS's comment, I came up with another idea to reintroduce the `#define print(sh, level, fmt, ...)` macro, but in a modified version that would be located in `net_shell_private.h`.
The name of the macro could be different, but it should belong to the `PR_*` family. I've named it `PR_S`, where 'S' stands for the safe check of the sh pointer.

```
#define PR_S(level, fmt, ...)						\
	do {								\
		if (sh) {						\
			shell_fprintf(sh, level, fmt, ##__VA_ARGS__); \
		} else {						\
			printk(fmt, ##__VA_ARGS__);			\
		}							\
	} while (false)

```

jukkar commented:
> I've named it `PR_S`, where 'S' stands for the safe check of the sh pointer.

No need to introduce new variants, the existing ones are enough.

EricNRS commented:
The `if` statement in the macro does increase the code size since there are two function calls in addition to the branch.

Take another look at my previous comment https://github.com/zephyrproject-rtos/zephyr/pull/68812#issuecomment-1936853333.  This would fix the problem as well without the extra call code overhead.  The `__ASSERT_NO_MSG(sh);` will ensure that there are not any existing use cases that expect sh to be NULL.

simensrostad reviewed (approved):
Resolves fatals seen when enabling shell in downstream nrf wifi samples

ndrs-pst committed:
wifi: shell: removed NULL check to net_mgmt callback

Since PR, PR_SHELL, PR_ERROR, PR_INFO, and PR_WARNING already have
an embedded `sh` NULL check, we can remove the change from PR #68809.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>
